backend/src/crawler/crawler_base.py

```diff
@@ -0,0 +1,147 @@
+    )
+    url: AnyHttpUrl | str = Field(
+        default=...,
+        example="https://www.example.com",
+        description="The URL of the article"
+    )
+
+
+class News(Headline):
+    time: str = Field(
+        default=...,
+        example="2021-10-01T00:00:00",
+        description="The time the article was published"
+    )
+    content: str = Field(
+        default=...,
+        example="Content of the article",
+        description="The content of the article"
+    )
+
+
+class NewsWithSummary(News):
+    summary: str = Field(
+        default=...,
+        example="Summary of the article",
+        description="The summary of the article"
+    )
+    reason: str = Field(
+        default=...,
+        example="Reason of the article",
+        description="The reason of the article"
+    )
+
+
+class NewsCrawlerBase(metaclass=abc.ABCMeta):
+    news_website_url: AnyHttpUrl | str
+    news_website_news_child_urls: list[AnyHttpUrl | str]
+
+    @abc.abstractmethod
+    def get_headline(
+            self, search_term: str, page: int | tuple[int, int]
+    ) -> list[Headline]:
+        """
+        Searches for news headlines on the news website based on a given search term and returns a list of headlines.
+
+        This method searches through the entire news_website_url using the specified search term, and returns a list
+        of Headline namedtuples, where each Headline includes the title and URL of a news article. The page parameter
+        can be an integer representing a single page number or a tuple representing a range of page numbers to search
+        through.
+        # The offset and limit parameters apply to the resulting list of headlines, allowing you to skip a
+        # certain number of headlines and limit the number of headlines returned, respectively.
+
+        :param search_term: A search term to search for news articles.
+        :param page: A page number (int) or a tuple of start and end page numbers (tuple[int, int]).
+        # :param offset: The number of headlines to skip from the beginning of the list.
+        # :param limit: The maximum number of headlines to return.
+        :return: A list of Headline namedtuple  s, each containing a title and a URL.
+        """
+        pass
+
+    @abc.abstractmethod
+    def parse(self, url: AnyHttpUrl | str) -> News:
+        """
+        Given a news URL from the news website, fetch and parse the detailed news content.
+
+        This method takes a URL that belongs to a news article on the news_website_url, retrieves the full content of
+        the news article, and returns it in the form of a News namedtuple. The News namedtuple includes the title,
+        URL, publication time, and content of the news article.
+
+        :param url: The URL of the news article to be fetched and parsed.
+        :return: A News namedtuple containing the title, URL, time, and content of the news article.
+        """
+        pass
```

Comment:
Since the comment stated that the method takes a news URL from the news website and this base class defines `validate_and_parse` to validate url and pass it to this method. I'd suggest it be renamed to `_parse`, which means the method is intended to only be used internally. Developers can call `validate_and_parse` to perform parsing news with url, preventing from passing invalid urls.